test/framework/src/main/java/org/elasticsearch/transport/nio/channel/TcpReadContext.java

```diff
@@ -0,0 +1,199 @@
+// selector.getBigArrays().newByteArray(messageLength), messageLength);
+//        BytesRef ref1 = ref.toBytesRef();
+//        ByteBuffer messageBuffer = ByteBuffer.wrap(ref1.bytes,ref1.offset, ref1.length - ref1.offset);
+
+        try {
+            if (headerBytesRemaining > 0) {
+                int headerStatus = readPartialHeader();
+                switch (headerStatus) {
+                    case PING:
+                        resetMessageStateMachine();
+                        handler.handlePing(channel, channel.getProfile());
+                        return bytesRead;
+                    case READ_INCOMPLETE:
+                        return bytesRead;
+                    case PEER_CLOSED:
+                        resetMessageStateMachine();
+                        return -1;
+                    case READ_COMPLETE:
+                }
+            }
+
+            int readStatus = readMessage();
+            switch (readStatus) {
+                case READ_INCOMPLETE:
+                    return bytesRead;
+                case PEER_CLOSED:
+                    resetMessageStateMachine();
+                    return -1;
+                case READ_COMPLETE:
+
+            }
+            BytesReference reference = new BytesArray(messageBuffer.array());
+
+            handler.handleMessage(reference, channel, channel.getProfile(), messageBytesLength);
+            resetMessageStateMachine();
+        } catch (Exception e) {
+            resetMessageStateMachine();
+            throw e;
+        }
+        return bytesRead;
+    }
+
+    private int readPartialHeader() throws IOException {
+        int headerBytesRead = channel.read(headerBuffer);
+        if (headerBytesRead < 0) {
+            return PEER_CLOSED;
+        }
+
+        bytesRead += headerBytesRead;
+        headerBytesRemaining -= headerBytesRead;
+        if (headerBytesRemaining != 0) {
+            return READ_INCOMPLETE;
+        }
+        int messageLength = readHeaderBuffer(headerBuffer);
+        if (messageLength == 0) {
+            return PING;
+        }
+        messageBytesLength = messageLength;
+        messageBytesRemaining = messageLength;
+        messageBuffer = ByteBuffer.allocate(messageLength);
+        return READ_COMPLETE;
+    }
+
+    private int readMessage() throws IOException {
+        int messageBytesRead = channel.read(messageBuffer);
+        if (messageBytesRead < 0) {
+            return PEER_CLOSED;
+        }
+
+        bytesRead += messageBytesRead;
+        messageBytesRemaining -= messageBytesRead;
+
+        if (messageBytesRemaining != 0) {
+            return READ_INCOMPLETE;
+        }
+        return READ_COMPLETE;
+    }
+
+    private void resetMessageStateMachine() {
+        headerBuffer.clear();
+        headerBytesRemaining = 6;
+        messageBytesLength = -1;
+        messageBytesRemaining = -1;
+    }
+
+    private int readHeaderBuffer(ByteBuffer headerBuffer) throws StreamCorruptedException {
```

Comment:
I wonder if we can extract and share this from TCP somehow.. can happen before or after this lands...